packages/compile-solidity/index.js

```diff
@@ -63,54 +87,76 @@ const Compile = {
     debug("compilationTargets: %O", compilationTargets);
 
-    // we can exit if there are no Solidity files to compile since
+    // we can exit if there are no Solidity/Yul files to compile since
     // it indicates that we only have Vyper-related JSON
     const solidityTargets = compilationTargets.filter(fileName =>
       fileName.endsWith(".sol")
     );
-    if (solidityTargets.length === 0) {
+    if (solidityTargets.length === 0 && yulPaths.length === 0) {
       return { compilations: [] };
     }
 
-    reportSources({ paths: compilationTargets, options });
+    reportSources({ paths: [...compilationTargets, ...yulPaths], options });
 
-    // when there are no sources, don't call run
-    if (Object.keys(allSources).length === 0) {
-      return { compilations: [] };
+    let solidityCompilations = [];
+    let yulCompilations = [];
+
+    // only call run if there are sources to run on!
+    if (Object.keys(allSources).length > 0) {
+      const solidityOptions = options.with({ compilationTargets });
+      debug("Compiling Solidity");
+      const compilation = await run(allSources, solidityOptions);
+      debug("Solidity compiled successfully");
+
+      // returns CompilerResult - see @truffle/compile-common
+      if (compilation.contracts.length > 0) {
+        solidityCompilations = [compilation];
+      }
     }
 
-    options.compilationTargets = compilationTargets;
-    const { sourceIndexes, sources, contracts, compiler } = await run(
-      allSources,
-      normalizeOptions(options)
-    );
+    for (const path of yulPaths) {
+      const yulOptions = options.with({ compilationTargets: [path] });
+      //load up Yul sources, since they weren't loaded up earlier
+      //(we'll just use FS for this rather than going through the resolver,
+      //for simplicity, since there are no imports to worry about)
+      const yulSource = fs.readFileSync(path, { encoding: "utf8" });
+      debug("Compiling Yul");
+      const compilation = await run({ [path]: yulSource }, yulOptions, "Yul");
+      debug("Yul compiled successfully");
+
+      // returns CompilerResult - see @truffle/compile-common
+      if (compilation.contracts.length > 0) {
+        yulCompilations.push(compilation);
+      }
+    }
+    if (yulPaths.length > 0 && !options.quiet) {
+      //replacement for individual Yul warnings
+      options.logger.log(
+        "> Warning: Yul is still experimental. Avoid using it in live deployments."
+      );
+    }
 
-    const { name, version } = compiler;
-    // returns CompilerResult - see @truffle/compile-common
-    return contracts.length > 0
-      ? {
-          compilations: [
-            {
-              sourceIndexes,
-              sources,
-              contracts,
-              compiler: { name, version }
-            }
-          ]
-        }
-      : { compilations: [] };
+    return {
+      compilations: [...solidityCompilations, ...yulCompilations]
+    };
   },
```

Comment:
OK hack is probably fine. Could we at least split this up into separate private methods or functions, one for Solidity specifics and one for Yul specifics?

Reply:
OK, I've factored out the Yul parts into their own function. I left the Solidity parts in place because that's both the original functionality and also much simpler -- it didn't seem worth factoring.  But obviously if you really think that should be pulled out as well it could be.